modAionImpl/src/org/aion/zero/impl/db/RecoveryUtils.java

```diff
@@ -364,4 +376,206 @@ public static void pruneOrRecoverState(String pruning_type) {
+        AionBlock block = store.getBestBlock();
+        AionBlock startBlock;
+        long currentBlock;
+        if (block != null && startHeight <= block.getNumber()) {
+            System.out.println(
+                    "\nImporting the main chain from block #"
+                            + startHeight
+                            + " to block #"
+                            + block.getNumber()
+                            + ". This may take a while.\n"
+                            + "The time estimates are optimistic based on current progress.\n"
+                            + "It is expected that later blocks take a longer time to import due to the increasing size of the database.\n");
+
+            if (startHeight == 0L) {
+                // dropping databases that can be inferred when starting from genesis
+                List<String> keep = List.of("block", "index");
+                repo.dropDatabasesExcept(keep);
+
+                // recover genesis
+                AionGenesis genesis = cfg.getGenesis();
+                AionHubUtils.buildGenesis(genesis, repo);
+                System.out.println("\nFinished rebuilding genesis block.");
+                startBlock = genesis;
+                currentBlock = 1L;
+            } else {
+                startBlock = store.getChainBlockByNumber(startHeight - 1);
+                currentBlock = startHeight;
+            }
+
+            boolean fail = false;
+
+            if (startBlock == null) {
+                System.out.println(
+                        "The main chain block at level "
+                                + currentBlock
+                                + " is missing from the database. Cannot continue importing stored blocks.");
+                fail = true;
+            } else {
+                chain.setBestBlock(startBlock);
+
+                long topBlockNumber = block.getNumber();
+                long stepSize = 10_000L;
+
+                Pair<ImportResult, AionBlockSummary> result;
+                final int THOUSAND_MS = 1000;
+
+                long start = System.currentTimeMillis();
+
+                // import in increments of 10k blocks
+                while (currentBlock <= topBlockNumber) {
+                    block = store.getChainBlockByNumber(currentBlock);
+                    if (block == null) {
+                        System.out.println(
+                                "The main chain block at level "
+                                        + currentBlock
+                                        + " is missing from the database. Cannot continue importing stored blocks.");
+                        fail = true;
+                        break;
+                    }
+
+                    try {
+                        result =
+                                chain.tryToConnectAndFetchSummary(
+                                        block, System.currentTimeMillis() / THOUSAND_MS, false);
```

Comment:
`TimeUnit` is a nice alternative to changing between time units; I don't really care either way though

Reply:
I'll make a note of this suggestion in the initial issue for later enhancement.